File: C#/Lab_23/Lab_23/Lab23_1.cs
using System;

namespace Lab_23
{
    public class Lab23_1
    {
        public abstract class Figure
        {
            public abstract double GetArea();
            public abstract double GetPerimeter();
        }

        public class Rectangle : Figure
        {
            public double Width, Height;

            public Rectangle(double w, double h)
            {
                Width = w;
                Height = h;
            }

            public override double GetArea() => Width * Height;
            public override double GetPerimeter() => 2 * (Width + Height);
        }

        public class Circle : Figure
        {
            public double Radius;

            public Circle(double r)
            {
                Radius = r;
            }

            public override double GetArea() => Math.PI * Radius * Radius;
            public override double GetPerimeter() => 2 * Math.PI * Radius;
        }

        public class Trapezium : Figure
        {
            public double A, B, C, D, Height;

            public Trapezium(double a, double b, double c, double d, double h)
            {
                A = a; B = b; C = c; D = d; Height = h;
            }

            public override double GetArea() => (A + B) / 2 * Height;
            public override double GetPerimeter() => A + B + C + D;
        }

        public interface IInterface<T>
        {
            void SetValue(T value);
            T GetValue();
        }

        public class MyClass<T> : IInterface<T>
        {
            private T value;

            public void SetValue(T value)
            {
                this.value = value;
            }

            public T GetValue()
            {
                return value;
            }
        }

        public static void Run()
        {
            Figure f1 = new Rectangle(5, 4);
            Figure f2 = new Circle(3);
            Figure f3 = new Trapezium(3, 5, 4, 4, 2);

            var obj = new MyClass<Figure>();

            obj.SetValue(f1);
            Console.WriteLine(obj.GetValue().GetArea());

            obj.SetValue(f2);
            Console.WriteLine(obj.GetValue().GetPerimeter());

            obj.SetValue(f3);
            Console.WriteLine(obj.GetValue().GetArea());
        }
    }
}

File: Java/Lab_23/src/Lab23_1.java
public class Lab23_1 {

    abstract static class Figure {
        abstract double getArea();
        abstract double getPerimeter();
    }

    static class Rectangle extends Figure {
        double width, height;

        Rectangle(double w, double h) {
            width = w;
            height = h;
        }

        double getArea() {
            return width * height;
        }

        double getPerimeter() {
            return 2 * (width + height);
        }
    }

    static class Circle extends Figure {
        double radius;

        Circle(double r) {
            radius = r;
        }

        double getArea() {
            return Math.PI * radius * radius;
        }

        double getPerimeter() {
            return 2 * Math.PI * radius;
        }
    }

    static class Trapezium extends Figure {
        double a, b, c, d, height;

        Trapezium(double a, double b, double c, double d, double h) {
            this.a = a;
            this.b = b;
            this.c = c;
            this.d = d;
            this.height = h;
        }

        double getArea() {
            return (a + b) / 2 * height;
        }

        double getPerimeter() {
            return a + b + c + d;
        }
    }

    interface IInterface<T> {
        void setValue(T value);
        T getValue();
    }

    static class MyClass<T> implements IInterface<T> {
        private T value;

        public void setValue(T value) {
            this.value = value;
        }

        public T getValue() {
            return value;
        }
    }

    public static void run() {
        Figure f1 = new Rectangle(5, 4);
        Figure f2 = new Circle(3);
        Figure f3 = new Trapezium(3, 5, 4, 4, 2);

        MyClass<Figure> obj = new MyClass<>();

        obj.setValue(f1);
        System.out.println(obj.getValue().getArea());

        obj.setValue(f2);
        System.out.println(obj.getValue().getPerimeter());

        obj.setValue(f3);
        System.out.println(obj.getValue().getArea());
    }
}
